pal03377/Athena-New
# Fix export script issues and add final evaluation data selection

> ⚠️ I recommend deleting the old evaluation data!

### Motivation and Context



I noticed some issues in the exported data when woking on #61 that needed fixing.

### Description


- [x] Use participation id instead of submission id
- [x] For a participation use the latest rated submission
- [x] For the feedback use the latest rated result of the submission
- [x] Fix off by one issues for text block indices for text exercise feedback
- [x] Fix off by one issue for the line number of programming feedback
- [x] Remove `.git` from repository export (not fully anonymized)
- [x] Better error handling for the repository download script with retries
- [x] Adjusted text exercises to use for evaluation (see screenshots below)
- [x] Change `content` to `text` after #47 is merged and add `grading_instruction_id`s
- [x] Add final selection of evaluation data ids for an admin to export
    - [x] Text exercises (I guess it is a lot right now, I just selected all exercises for now, we can eliminate later)
    - [x] Programming exercises (has to be a small set of exercises because the courses are very large and the repositories are rather big)
- [x] Add more control what is being downloaded
    - Select what to download starting of without a selection (so that we don't need to redownload everything)
> ⚠️ The export feels much slower now but should at least be correct now

### Steps for Testing


1. cd playground
2. Have the anonymized database dump loaded 
3. `npm run export:artemis:2-export-evaluation-data`
4. Script should not error or abort/timeout (Probably takes now about a minute or two 🤔)
5. Put the json data from below in the `playground/scripts/artemis/evaluation_data/programming_exercises.json`
6. Test the repository export error handling: `npm run export:artemis:3-download-programming-repositories`
   1. Use default Artemis production url and just use `artemis_test_user_11`
   2. Just press `Y` when asked to download the (iPraktikum) exercises
   3. `artemis_test_user_11` does not have access, it should retry and log the errors
   4. Check in the `playground/evaluation/exercise-10657/` dir that the materials and submissions do not contain the `.git` folder

> Note: `export:artemis:4-link-programming-repositories` does not need testing

### Example execution + Testing Data
<img width="1093" alt="image" src="https://github.com/ls1intum/Athena/assets/5898705/740c9fd2-726f-4bb5-b40c-56b5d18029ee">

Data used for testing on production with `artemis_test_user_15`:
```
{
  "Test": {
    "course_id": 10,
    "semester": "WS22/23",
    "exercises": [
      {
        "id": 6685,
        "title": "StudentCodeSeparator Experiments Daniel Kirschten"
      },
      {
        "id": 6835,
        "title": "StudentCodeSeparator Experiments Daniel Kirschten 2"
      }
    ]
  },
  "Test: Exam Mode Testing SS23": {
    "course_id": 10,
    "semester": "WS22/23",
    "exercises": [
      {	
        "id": 10573,
        "title": "normale prog ex test"
      },
      {	
        "id": 10576,
        "title": "sdfghshdf"
      },
      {	
        "id": 10657,
        "title": "Test Exercise"
      },
      {	
        "id": 10573,
        "title": "normale prog ex test"
      },
      {	
        "id": 10576,
        "title": "sdfghshdf"
      },
      {	
        "id": 10657,
        "title": "Test Exercise"
      },
      {	
        "id": 10573,
        "title": "normale prog ex test"
      },
      {	
        "id": 10576,
        "title": "sdfghshdf"
      },
      {	
        "id": 10657,
        "title": "Test Exercise"
      }
    ]
  }
}
```

### Screenshots

**Jan-Philip's evaluation/validation courses:**
<img width="685" alt="image" src="https://github.com/ls1intum/Athena/assets/5898705/2b771a4d-41b0-4481-9246-a6cf530135dd">

**Most of the courses have a lot of empty feedback (i.e. `grading_instruction_id` probably was removed):**
<img width="401" alt="image" src="https://github.com/ls1intum/Athena/assets/5898705/a46b7902-7c1d-4cd1-a4a4-6f1e30524ddc">
I found only ISE SS21 to be usable (if we want feedback data) and 3 more potential courses we could evaluate on, all highlighted in grey.

**For programming exercises I just picked what looked like good data:**
Two exercises per programming language (C, Assembly, VHDL, OCaml Haskel, Python) and 5 for Java. I tried to pick some with the most manual feedback which ended up mostly being exam exercises.

<img width="1408" alt="image" src="https://github.com/ls1intum/Athena/assets/5898705/27e5c17a-7754-4564-a219-04c0b75b7829">
